### PR TITLE
#663: Fixed bitcount operations

### LINKED ISSUE #663: Inconsistent `BITCOUNT`: BITCOUNT fuzzing without start/end
## Steps to reproduce

Run the commands mentioned in the test on Line 114 in the file https://github.com/AshwinKul28/dice-tests/blob/main/tcltests/unit/bitpos.tcl#L114.

## Expected output

The expected output when the above set of commands when run on Redis

```
Expected [r bitcount str 0 -1 bit]
```

## Observed output

The observed output when the above set of commands when run on DiceDB

```
2153 (context: type eval line 7 cmd {assert {[r bitcount str 0 -1 bit]
```

The steps to run the test cases are mentioned in the README of the [dice-tests repository](https://github.com/AshwinKul28/dice-tests).

## Expectations for resolution

This issue will be considered resolved when the following things are done:

1. Changes in the [`dice`](https://github.com/dicedb/dice) repository code to meet the expected behavior.
1. Successful run of the tcl test behavior.

You can find the tests under the `tests` directory of the [`dice`](https://github.com/dicedb/dice) repository and the steps to run are in the [README file](https://github.com/dicedb/dice). Refer to the following links to set up DiceDB and Redis 7.2.5 locally

- [Setup DiceDB locally](https://github.com/dicedb/dice)
- [Setup Redis 7.2.5 locally](https://gist.github.com/arpitbbhayani/94aedf279349303ed7394197976b6843)
- [Setup DiceDB CLI](https://github.com/dicedb/dice)

### PR DESCRIPTION
**Pull request by VedWhat**
This PR attempts to close #663 however I realised rewrite would be a better option. This passes all tests under tcl (only for bitcount) and all integration tests. While this also fixes other issues, I don't want to overstep.

I've added comments wherever necessary but can elaborate wherever necessary

What has changed?
* Added better argument validation
* Changed counting in a loop to count using a bitmask
* Added better handling of start and end indices.

@lucifercr07 as mentioned, this does fix the other bitcount issues, lmk what you think. I believe this is a better way to fix the errors (fixing a single test would've made it messier if you get what I mean)

Happy to change the approach just to fix the single test.

### CODE REVIEW

**Review by lucifercr07 — APPROVED** [review]
LGTM, it's more cleaner than prior implementation.
@JyotinderSingh shall we merge this?

@VedWhat on other note what are your thoughts on using go inbuilt function i.e. [OnesCount8](https://pkg.go.dev/math/bits#OnesCount8) rather than the `popcount` implementation?

File: internal/eval/eval.go [hunk 1]
```diff
@@ -2039,106 +2039,115 @@ func evalBITCOUNT(args []string, store *dstore.Store) []byte {
...
-		return clientio.Encode(0, true)
-	}
-	if end > valueLength && unit == BYTE {
-		end = valueLength - 1
 	}
 
-	bitCount := 0
-	if unit == BYTE {
+	switch unit {
+	case "BYTE":
```
Comment (JyotinderSingh) [inline internal/eval/eval.go:2089]: Let's use a constant for this

File: internal/eval/eval.go [hunk 2]
```diff
@@ -2039,106 +2039,115 @@ func evalBITCOUNT(args []string, store *dstore.Store) []byte {
...
-			considerBits := start % 8
-			for j := 8 - considerBits - 1; j >= 0; j-- {
-				bitCount += int(popcount(byte(int(value[i]) & (1 << j))))
-			}
-		} else if i == endBitRange {
-			considerBits := end % 8
-			for j := considerBits; j >= 0; j-- {
-				bitCount += int(popcount(byte(int(value[i]) & (1 << (8 - j - 1)))))
-			}
+	case "BIT":
```
Comment (JyotinderSingh) [inline internal/eval/eval.go:2105]: Let's use a constant for this

**Review by JyotinderSingh — APPROVED** [review]
LGTM. Few minor comments.

**Comment by VedWhat** [comment]
> LGTM, it's more cleaner than prior implementation. @JyotinderSingh shall we merge this?
> 
> @VedWhat on other note what are your thoughts on using go inbuilt function i.e. [OnesCount8](https://pkg.go.dev/math/bits#OnesCount8) rather than the `popcount` implementation?

Nice, makes sense to use a lookup instead of any operations. Made the change!

**Comment by VedWhat** [comment]
> > > LGTM, it's more cleaner than prior implementation. @JyotinderSingh shall we merge this?
> > 
> > 
> > > 
> > 
> > 
> > > @VedWhat on other note what are your thoughts on using go inbuilt function i.e. [OnesCount8](https://pkg.go.dev/math/bits#OnesCount8) rather than the `popcount` implementation?
> > 
> > 
> > Nice, makes sense to use a lookup instead of any operations. Made the change!
> 
> thanks, could you take a look at the linter issues too?

right, was pushing with a no-verify due to an issue. already pushed the change.

**Pull request merged**